presto-main/src/main/java/io/prestosql/server/remotetask/DynamicFiltersFetcher.java

```diff
@@ -0,0 +1,239 @@
+    private final JsonCodec<VersionedDynamicFilterDomains> dynamicFilterDomainsCodec;
+    private final Duration refreshMaxWait;
+    private final Executor executor;
+    private final HttpClient httpClient;
+    private final RequestErrorTracker errorTracker;
+    private final RemoteTaskStats stats;
+    private final AtomicLong currentRequestStartNanos = new AtomicLong();
+
+    @GuardedBy("this")
+    private long dynamicFiltersVersion = INITIAL_DYNAMIC_FILTERS_VERSION;
+    @GuardedBy("this")
+    private long localDynamicFiltersVersion;
+    @GuardedBy("this")
+    private Map<DynamicFilterId, Domain> dynamicFilterDomains = ImmutableMap.of();
+    @GuardedBy("this")
+    private boolean running;
+    @GuardedBy("this")
+    private ListenableFuture<FullJsonResponseHandler.JsonResponse<VersionedDynamicFilterDomains>> future;
+
+    public DynamicFiltersFetcher(
+            Consumer<Throwable> onFail,
+            TaskId taskId,
+            URI taskUri,
+            Duration refreshMaxWait,
+            JsonCodec<VersionedDynamicFilterDomains> dynamicFilterDomainsCodec,
+            Executor executor,
+            HttpClient httpClient,
+            Duration maxErrorDuration,
+            ScheduledExecutorService errorScheduledExecutor,
+            RemoteTaskStats stats)
+    {
+        this.taskId = requireNonNull(taskId, "taskId is null");
+        this.taskUri = requireNonNull(taskUri, "taskUri is null");
+        this.onFail = requireNonNull(onFail, "onFail is null");
+
+        this.refreshMaxWait = requireNonNull(refreshMaxWait, "refreshMaxWait is null");
+        this.dynamicFilterDomainsCodec = requireNonNull(dynamicFilterDomainsCodec, "dynamicFilterDomainsCodec is null");
+
+        this.executor = requireNonNull(executor, "executor is null");
+        this.httpClient = requireNonNull(httpClient, "httpClient is null");
+
+        this.errorTracker = new RequestErrorTracker(taskId, taskUri, maxErrorDuration, errorScheduledExecutor, "getting dynamic filter domains");
+        this.stats = requireNonNull(stats, "stats is null");
+    }
+
+    public synchronized void start()
+    {
+        if (running) {
+            // already running
+            return;
+        }
+        running = true;
+        scheduleNextRequestIfNecessary();
+    }
+
+    public synchronized void stop()
+    {
+        running = false;
+        if (future != null) {
+            future.cancel(true);
+            future = null;
+        }
+    }
+
+    public synchronized Map<DynamicFilterId, Domain> getDynamicFilterDomains()
+    {
+        return dynamicFilterDomains;
+    }
+
+    public synchronized void updateDynamicFiltersVersion(long newDynamicFiltersVersion)
+    {
+        if (dynamicFiltersVersion >= newDynamicFiltersVersion) {
+            return;
+        }
+
+        dynamicFiltersVersion = newDynamicFiltersVersion;
+        scheduleNextRequestIfNecessary();
+    }
+
+    private synchronized void scheduleNextRequestIfNecessary()
```

Comment:
I would name this "fetchDynamicFilterIfNecessary".. This name is used for the continuous fetchers, so seems a little misleading here.